File: ros2_ws/src/rcdt_sensors/src_py/combine_camera_topics.py
#!/usr/bin/env python3

# SPDX-FileCopyrightText: Alliander N. V.
#
# SPDX-License-Identifier: Apache-2.0

import rclpy
from rcdt_utilities.launch_utils import spin_node
from rclpy import logging
from rclpy.node import Node
from realsense2_camera_msgs.msg import RGBD
from sensor_msgs.msg import CameraInfo, Image

ros_logger = logging.get_logger(__name__)


class CombineCameraTopics(Node):
    """A ROS2 node that combines RGB and depth camera topics into a single RGBD message.

    This node subscribes to the RGB and depth image topics, as well as their corresponding
    camera info topics. It publishes a combined RGBD message that contains both the RGB image
    and the depth image, along with their camera information.
    """

    def __init__(self) -> None:
        """Initialize the CombineCameraTopics node."""
        super().__init__("combine_camera_topics")
        namespace = self.get_namespace().strip("/")

        self.declare_parameter("rgb", f"/{namespace}/color/image_raw")
        self.declare_parameter("depth", f"/{namespace}/depth/image_rect_raw")
        self.declare_parameter("rgb_info", f"/{namespace}/color/camera_info")
        self.declare_parameter("depth_info", f"/{namespace}/depth/camera_info")
        self.declare_parameter("rgbd", f"/{namespace}/rgbd")

        rgb = self.get_parameter("rgb").get_parameter_value().string_value
        depth = self.get_parameter("depth").get_parameter_value().string_value
        rgb_info = self.get_parameter("rgb_info").get_parameter_value().string_value
        depth_info = self.get_parameter("depth_info").get_parameter_value().string_value
        rgbd = self.get_parameter("rgbd").get_parameter_value().string_value

        self.create_subscription(Image, rgb, self.update_rgb, 10)
        self.create_subscription(Image, depth, self.update_depth, 10)
        self.create_subscription(CameraInfo, rgb_info, self.update_rgb_info, 10)
        self.create_subscription(CameraInfo, depth_info, self.update_depth_info, 10)

        self.rgbd = RGBD()
        self.pub = self.create_publisher(RGBD, rgbd, 10)
        self.create_timer(1 / 30, self.publish_rgbd)

    def publish_rgbd(self) -> None:
        """Publish the combined RGBD message."""
        self.pub.publish(self.rgbd)

    def update_rgb(self, msg: Image) -> None:
        """Update the RGB image in the RGBD message.

        Args:
            msg (Image): The incoming RGB image message.
        """
        self.rgbd.rgb = msg

    def update_depth(self, msg: Image) -> None:
        """Update the depth image in the RGBD message.

        Args:
            msg (Image): The incoming depth image message.
        """
        self.rgbd.depth = msg

    def update_rgb_info(self, msg: CameraInfo) -> None:
        """Update the RGB camera info in the RGBD message.

        Args:
            msg (CameraInfo): The incoming RGB camera info message.
        """
        self.rgbd.rgb_camera_info = msg

    def update_depth_info(self, msg: CameraInfo) -> None:
        """Update the depth camera info in the RGBD message.

        Args:
            msg (CameraInfo): The incoming depth camera info message.
        """
        self.rgbd.depth_camera_info = msg


def main(args: list | None = None) -> None:
    """Main function to initialize the ROS 2 node and start spinning.

    Args:
        args (list | None): Command line arguments, defaults to None.
    """
    rclpy.init(args=args)
    node = CombineCameraTopics()
    spin_node(node)


if __name__ == "__main__":
    main()

File: ros2_ws/src/rcdt_sensors/src/combine_camera_topics.cpp
// # SPDX-FileCopyrightText: Alliander N. V.
//
// # SPDX-License-Identifier: Apache-2.0

#include "combine_camera_topics.hpp"
#include <chrono>
#include <rclcpp/executors.hpp>
#include <rclcpp/parameter_value.hpp>

using std::placeholders::_1;

CombineCameraTopics::CombineCameraTopics() : Node("combine_camera_topics") {

  this->declare_parameter("rgb_topic", "/camera/camera/color/image_raw");
  this->declare_parameter("depth_topic", "/camera/camera/depth/image_rect_raw");
  this->declare_parameter("rgb_info_topic", "/camera/camera/color/camera_info");
  this->declare_parameter("depth_info_topic",
                          "/camera/camera/depth/camera_info");
  this->declare_parameter("rgbd_topic", "/camera/camera/rgbd");
  this->declare_parameter("rate", 30);

  initialize_subscriptions();
  initialize_publisher();
};

void CombineCameraTopics::initialize_subscriptions() {
  auto rgb_topic = this->get_parameter("rgb_topic").as_string();
  sub_rgb = this->create_subscription<Image>(
      rgb_topic, 10, std::bind(&CombineCameraTopics::rgb_callback, this, _1));
  auto depth_topic = this->get_parameter("depth_topic").as_string();
  sub_depth = this->create_subscription<Image>(
      depth_topic, 10,
      std::bind(&CombineCameraTopics::depth_callback, this, _1));
  auto rgb_info_topic = this->get_parameter("rgb_info_topic").as_string();
  sub_rgb_info = this->create_subscription<CameraInfo>(
      rgb_info_topic, 10,
      std::bind(&CombineCameraTopics::rgb_info_callback, this, _1));
  auto depth_info_topic = this->get_parameter("depth_info_topic").as_string();
  sub_depth_info = this->create_subscription<CameraInfo>(
      depth_info_topic, 10,
      std::bind(&CombineCameraTopics::depth_info_callback, this, _1));
}

void CombineCameraTopics::initialize_publisher() {
  auto rgbd_topic = this->get_parameter("rgbd_topic").as_string();
  pub_rgbd = this->create_publisher<RGBD>(rgbd_topic, 10);
  auto rate = this->get_parameter("rate").as_int();
  timer = this->create_wall_timer(
      std::chrono::duration<double>(1.0 / rate),
      std::bind(&CombineCameraTopics::timer_callback, this));
}

void CombineCameraTopics::rgb_callback(Image msg) { rgbd.rgb = msg; }
void CombineCameraTopics::depth_callback(Image msg) { rgbd.depth = msg; }
void CombineCameraTopics::rgb_info_callback(CameraInfo msg) {
  rgbd.rgb_camera_info = msg;
}
void CombineCameraTopics::depth_info_callback(CameraInfo msg) {
  rgbd.depth_camera_info = msg;
}

void CombineCameraTopics::timer_callback() { pub_rgbd->publish(rgbd); };

int main(int argc, char **argv) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<CombineCameraTopics>());
  rclcpp::shutdown();
}